[JIT] Factor Module into Object and Module

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29501 [WIP][JIT] Separate torchbind from Python
* **#29500 [JIT] Factor Module into Object and Module**

Differential Revision: [D18463064](https://our.internmc.facebook.com/intern/diff/D18463064)

Commit:
[JIT] Factor Module into Object and Module

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/script/module.h:
```diff
@@ -125,17 +126,117 @@ struct TORCH_API Method {
 };
 
-struct TORCH_API Module {
+struct TORCH_API Object {
```
Object should go in its own header and compilation unit. It will grow over time, and it is confusing for it to be in module.h

zdevito commented on torch/csrc/jit/script/module.h:
```diff
@@ -125,17 +126,117 @@ struct TORCH_API Method {
+    object_value()->setSlot(slot, std::move(v));
+  }
+
+  ObjectPtr object_value() const;
```
Since we are going to go through the effort of renaming `module_value`, I think this should be called `_ivalue`. In the future, we are not going to expose this publically.

zdevito commented on torch/csrc/jit/script/module.cpp:
```diff
@@ -14,7 +14,7 @@ namespace torch {
 namespace script {
 
-static ModulePtr create_module_object(
+static ObjectPtr create_object_value(
```
This looks like a mistaken find/replace renaming. Should still be called `create_module_object` given how it is used.

zdevito commented on torch/csrc/jit/script/module.cpp:
```diff
@@ -33,34 +33,34 @@ static ModulePtr create_module_object(
-ModulePtr Module::module_object() const {
-  if (!module_value_) {
+ObjectPtr Object::object_value() const {
+  if (!object_value_) {
```
The behavior in here is specific to Module and not object -- only Module are used in tests this way, and this will break most uses of object. We probably should go through and find the places where `Module m;` is meant to create a blank module, and change it to an explicit call to create a blank module. That way, this code can just assert that `object_value_` exists.

zdevito commented on torch/csrc/jit/script/module.h:
```diff
@@ -125,17 +126,117 @@ struct TORCH_API Method {
+  Object() {}
+  Object(ObjectPtr object_value) : object_value_(std::move(object_value)) {}
+
+  const c10::QualifiedName& name() const {
```
We should kill this method and replace it with `obj.type()->name()` in its uses. Asking for the name of an object doesn't really make sense.

zdevito commented on torch/csrc/jit/script/module.h:
```diff
@@ -125,17 +126,117 @@ struct TORCH_API Method {
+  ClassTypePtr type() const {
+    return object_value()->type();
+  }
+  std::shared_ptr<CompilationUnit> class_compilation_unit() const {
```
Let's not bind this publically. If there are remaining public uses then object_value()->compilation_unit() can replace it.

zdevito commented on torch/csrc/jit/script/module.h:
```diff
@@ -125,17 +126,117 @@ struct TORCH_API Method {
+  void define(const std::string& src, const ResolverPtr& resolver = nullptr);
+
+  template <typename... Types>
+  IValue create_class(const c10::QualifiedName& name, Types&&... args) const {
```
We shouldn't bind `create_class` into Object. First it probably should be called `create_object` because it is not creating a new type. Second it is really only creating an object in the same compilation unit as this object. I am not sure that should be a method on Object. We can keep these calls on Module until we have a better place for them.

zdevito reviewed (commented):
Thanks for doing this refactor. I have inline comments about details, but they should not be that complicated to address.

Commit:
Update on "[JIT] Factor Module into Object and Module"




[ghstack-poisoned]

Commit:
Update on "[JIT] Factor Module into Object and Module"




[ghstack-poisoned]

Commit:
Update on "[JIT] Factor Module into Object and Module"




[ghstack-poisoned]

Commit:
Update on "[JIT] Factor Module into Object and Module"


Differential Revision: [D18463064](https://our.internmc.facebook.com/intern/diff/D18463064)

[ghstack-poisoned]

Commit:
Update on "[JIT] Factor Module into Object and Module"


Differential Revision: [D18463064](https://our.internmc.facebook.com/intern/diff/D18463064)

[ghstack-poisoned]

Commit:
Update on "[JIT] Factor Module into Object and Module"


Differential Revision: [D18463064](https://our.internmc.facebook.com/intern/diff/D18463064)

[ghstack-poisoned]

Commit:
Update on "[JIT] Factor Module into Object and Module"


Differential Revision: [D18463064](https://our.internmc.facebook.com/intern/diff/D18463064)

[ghstack-poisoned]

Commit:
Update on "[JIT] Factor Module into Object and Module"


Differential Revision: [D18463064](https://our.internmc.facebook.com/intern/diff/D18463064)

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -532,10 +533,60 @@ void initJitScriptBindings(PyObject* module) {
+              return method.name();
+            });
+          })
+      .def_property_readonly("name", [](const Module& self) {
```
This one is similar to the C++ binding -- have a name property on an object is weird, it should probably be `._type.name()`a

zdevito reviewed (approved):
Looks good -- thanks!

Commit:
Update on "[JIT] Factor Module into Object and Module"


Differential Revision: [D18463064](https://our.internmc.facebook.com/intern/diff/D18463064)

[ghstack-poisoned]

Commit:
Update on "[JIT] Factor Module into Object and Module"


Differential Revision: [D18463064](https://our.internmc.facebook.com/intern/diff/D18463064)

[ghstack-poisoned]

Commit:
Update on "[JIT] Factor Module into Object and Module"


Differential Revision: [D18463064](https://our.internmc.facebook.com/intern/diff/D18463064)

[ghstack-poisoned]

Commit:
Update on "[JIT] Factor Module into Object and Module"


Differential Revision: [D18463064](https://our.internmc.facebook.com/intern/diff/D18463064)

[ghstack-poisoned]

Commit:
Update on "[JIT] Factor Module into Object and Module"


Differential Revision: [D18463064](https://our.internmc.facebook.com/intern/diff/D18463064)

[ghstack-poisoned]

Commit:
Update on "[JIT] Factor Module into Object and Module"


Differential Revision: [D18463064](https://our.internmc.facebook.com/intern/diff/D18463064)

[ghstack-poisoned]

Commit:
Update on "[JIT] Factor Module into Object and Module"


Differential Revision: [D18463064](https://our.internmc.facebook.com/intern/diff/D18463064)

[ghstack-poisoned]

Commit:
Update on "[JIT] Factor Module into Object and Module"


Differential Revision: [D18463064](https://our.internmc.facebook.com/intern/diff/D18463064)

[ghstack-poisoned]

Commit:
Update on "[JIT] Factor Module into Object and Module"


Differential Revision: [D18463064](https://our.internmc.facebook.com/intern/diff/D18463064)

[ghstack-poisoned]
